processor/src/test/java/org/mapstruct/ap/test/accessingmappers/InventoryMapper.java

```diff
@@ -0,0 +1,49 @@
+/**
+ *  Copyright 2012-2016 Gunnar Morling (http://www.gunnarmorling.de/)
+ *  and/or other contributors as indicated by the @authors tag. See the
+ *  copyright.txt file in the distribution for a full listing of all
+ *  contributors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.mapstruct.ap.test.accessingmappers;
+
+import org.mapstruct.Mapper;
+import org.mapstruct.factory.Mappers;
+
+
+/**
+ *
+ * @author Sjaak Derksen
+ */
+@Mapper(uses = { RadioMapper.class, TvMapper.class } )
+public abstract class InventoryMapper {
+
+    public static final InventoryMapper INSTANCE = Mappers.getMapper( InventoryMapper.class );
+
+    private final InventorySequence inventorySequence = new InventorySequence();
+
+    public  InventoryEntity mapInventory(InventoryDto source) {
+        getRadioMapper().init( inventorySequence );
```

Comment:
The thing is that we kind of encourage static singletons via the INSTANCE pattern we often use.
  Ok.. I get it. You mean by assigning it to a final static we imply that Mappers are stateless. 
  
  To eleborate a bit more.. I've got a requirement that id's in a result document are generated unique within the scope of a soap request. Some of my mappers are shared between products as common utils. I don't want to revert to CDI only to produce mappers with state added at construction time. Running through the returned document the first entry I find get 1, the next 2, etc. ID's are required for common java / XSD classes. I now use a thread local to get the ID generater (basically a wrapped counter) to the right mapper (generated, but also hand written). But I don't like thread-locals and try to avoid them were possible. Using a classic singleton patterns makes the problem only worse, since then my sequence generated is shared between all threads running in my VM.
  Anyway. I don't think MapStruct should impose limitatons on how a user would like to configure a mapper (completely hand written or generated).
  A colleague of mine stumbled on another situation. The ones with QualityRegimes. He would like to configure the behaviour of a mapper depending based on the regime. Regime 'A' supports another set of string based value mappings then regime 'B'. Depending on the regime a different hash map is chosen when callign a qualifying method.
  Hope this makes the case a bit more clear.
  BTW: another option would be to support a factory for creation of Mappers. I reckon that impact will be quite large.
  
  
  
   Op 25 februari 2016 om 22:46 schreef Gunnar Morling <notifications@github.com>:
   
   
   In processor/src/test/java/org/mapstruct/ap/test/accessingmappers/InventoryMapper.java:
   > +import org.mapstruct.factory.Mappers;
> +
> +
> +/**
> + *
> + * @author Sjaak Derksen
> + */
> +@Mapper(uses = { RadioMapper.class, TvMapper.class } )
> +public abstract class InventoryMapper {
> +
> +    public static final InventoryMapper INSTANCE = Mappers.getMapper( InventoryMapper.class );
> +
> +    private final InventorySequence inventorySequence = new InventorySequence();
> +
> +    public  InventoryEntity mapInventory(InventoryDto source) {
> +        getRadioMapper().init( inventorySequence );

   The thing is that we kind of encourage static singletons via the INSTANCE pattern we often use. How about making your sequence a singleton, obtainable via a static method such as Sequence#getInstance()?
   —Reply to this email directly or view it on GitHub.
   
     
    
   
  
   